### PR TITLE
KAFKA-3930: Replace colon in ipv6 addresses used in metrics tags

### PR DESCRIPTION
**Pull request by rajinisivaram**

### CODE REVIEW

File: core/src/main/scala/kafka/common/ClientIdAndBroker.scala [hunk 1]
```diff
@@ -26,6 +26,10 @@ trait ClientIdBroker {
 }
 
 case class ClientIdAndBroker(clientId: String, brokerHost: String, brokerPort: Int) extends ClientIdBroker {
+
+  // ipv6 addresses contain ':' which is not a valid character in metrics ObjectName
+  def brokerHostMetricsTag = brokerHost.replace(':', '.')
```
Comment (granthenke) [inline core/src/main/scala/kafka/common/ClientIdAndBroker.scala]: I think we do this type of "sanitizing" for multiple metrics and I think its a standard pattern that we would do for any colon (':'). Maybe we should move this into the code for defining a metric. Somewhere under the `newMeter` implementation? It looks like `KafkaMetricsGroup.metricName` already tries to return a "Sanitized metric name object." Maybe that would be the place?

Comment (rajinisivaram) [reply core/src/main/scala/kafka/common/ClientIdAndBroker.scala]: @granthenke Thank you for the review. I was thinking earlier that perhaps we dont want to change a tag that contained colon by mistake. But I suppose it is better to change any occurrence of colon, making the metric name valid. I have moved the code to `KafkaMetricsGroup`.

**Comment by harshach** [comment]
+1

**Comment by hachikuji** [comment]
@rajinisivaram I'm wondering why we don't use `ObjectName.quote` to make the tag values safe generically. This is what is done internally in the Yammer `MetricName.createMBeanName` method.

**Comment by rajinisivaram** [comment]
@hachikuji Thank you for the review. Makes sense to quote tags instead of replacing :, have updated the code to do that. The code adds quotes when necessary (ipv6 address for instance), leaving other tags unchanged.

**Comment by hachikuji** [comment]
Hey @rajinisivaram, this LGTM for the specific problem. I think it would be worth adding a similar fix in `org.apache.kafka.common.metrics.JmxReporter`. Although we don't appear to include the hostname for metrics in `NetworkClient`, it seems like we probably should (I'll create a separate JIRA for this).

**Comment by hachikuji** [comment]
Thanks for the patch! Merging to trunk and 0.10.1 since this fixes the reported problem. I've created [KAFKA-4243](https://issues.apache.org/jira/browse/KAFKA-4243) for more general changes.

**Pull request closed without merging**